SO/Ficha5.md

```diff
@@ -0,0 +1,203 @@
+
+Quando a função bloqueia, fica num estado suspenso, à espera de uma de duas coisas, 
+ou que outro processo escreva no extremo de escrita, dando à função algo para ler,
+ou que o extremo de escrita seja fechado em todos os outros processos com acesso ao pipe.
+
+#### Escrita
+
+A tabela seguinte ilustra o comportamento da função `write(pipefd[1], buffer, BUFFERSIZE)`,
+onde `pipefd[1]` é o extremo de escrita do pipe, tal como foi definido no exemplo acima.
+
+| Extremo de leitura está | Há espaço | Não há espaço |
+|-------------------|-----|-----|
+| Fechado | Mata escritor (SIGPIPE) | Mata escritor (SIGPIPE) |
+| Aberto | Escreve do buffer para o pipe | Bloqueia |
+
+O sinal `SIGPIPE` indica que tentámos escrever num pipe cujos extremos de leitura
+estão fechados. Isto diz-nos que, para podermos escrever num pipe, este tem que
+ter pelo menos um extremo de leitura aberto.
+
+#### Exemplo de leitura e escrita num pipe
+
+Neste exemplo, um processo-pai envia a um processo filho a mensagem "Bom dia",
+que a imprime no terminal.
+
+```C
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/wait.h>
+
+int main() {
+    int pipefd[2];
+    if(pipe(pipefd) < 0) {
+        perror("Erro na criação do pipe!");
+        exit(1);
+    }
+
+    if(fork() == 0) { // Processo-filho
+
+        // O processo-filho não vai escrever no pipe, portanto podemos fechar o extremo de escrita.
+        close(pipefd[1]);
+
+        char buf[10];
+        int bytes_read = read(pipefd[0], buf, 10); // Estamos a ler do extremo de leitura do pipe.
+        close(pipefd[0]); // Podemos fechar o extremo de leitura pois já não precisamos de ler do pipe.
+
+        write(STDOUT_FILENO, buf, bytes_read);
+        _exit(0);
+    } 
+
+    else { // Processo-pai
+
+        // O processo-pai não vai ler do pipe, portanto podemos fechar o extremo de leitura.
+        close(pipefd[0]);
+
+        char * str = "Bom dia";
+        write(pipefd[1], str, strlen(str)); // Estamos a escrever para o extremo de escrita do pipe.
+        close(pipefd[1]); // Podemos fechar o extremo de leitura pois não vamos voltar a escrever no pipe.
+
+        wait(NULL);
+    }
+
+    return 0;
+}
+```
+
+O *output* deste programa será:
+> Bom dia
+
+## Pipes para encadeamento de comandos
+
+Um dos principais usos de pipes anónimos é no encadeamento de comandos. Se tivermos
+um ficheiro `foo.txt` que contém:
+```
+Bom dia!
+Hoje está um dia agradável.
+Tenho que estudar Sistemas Operativos.
+```
+e quisermos obter o número de linhas neste ficheiro, podemos usamos o seguinte comando:
+```bash
+$ cat foo.txt | wc -l
+```
+que, na verdade, é a combinação de dois comandos, um que dá como *output* o conteúdo
+do ficheiro `foo.txt` e outro que determina o número de linhas do *input* fornecido.
+
+Aquele `|` permite-nos usar o *output* de um comando como *input* de outro. É 
+possível emular esse comportamento em C com o uso de pipes.
```

Comment:
Nao e "emular" se for o mesmo comportamento
Acho que a frase devia ser: A shell implementa isto com pipes